### DISCUSSION TITLE
Help: many popups bug when opening the same panel twice

### DISCUSSION THREAD

**Discussion by dobo365** [RomanKubiak/ctrlr#200] [Q&A] [answered]
OK, I have this issue since long but I'd like to understand what's going on.
Before raising a Ctrlr issue, I'd like to be sure it is not my fault.
So hoping that some of you will be able to help.

**Issue:**
- using 5.3.201. Didn't try on others but will try with 5.6 and 6.0.26 to check.
- When you open the panel everything's looks nice and OnModulatorValueChange methods are not fired due to a isPanelReady() function (in Miscellaneous method). After opening, you can use Rename or Load as normal.
- When you open the panel a second time (or a diffirent panel but using the same philosophy and variable/method names), all popups (from the first panel) are fired while they should not
- When the panel is opened as exported VST then put on 2 tracks in the DAW, the same is occurring
- It is happening with some panels and not with others built the same way which is even more intriguing... (I'll try cleaning a panel that doesn't have it and bring it to the same minimal content so maybe I can see something)

**Reasoning:**
- is it due to utils.xxx ?
- is it due to the fact that global variables are used (bPanelLoaded) and that they are global for the whole Ctrlr and not by panel instance?
- it is not due to the Unique ID because I'm getting the same issue when opening different panels but using the same logic and variable names. I will try making the bPanelLoaded variable name unique and the isPanelReady() method also unique to check if better.

Here is a panel where I removed everything but what is giving the issue.
[Many popups bug - Cleaned.zip](https://github.com/RomanKubiak/ctrlr/files/6074660/Many.popups.bug.-.Cleaned.zip)
Instead of getting 2 popups, with a full panel, I'm getting more than 20...
So, must be solved/understood.

**Any help understanding the issue OR having a workaround would be much appreciated!**

Thx in advance for your help!

**Comment by Ted-juh** [comment]
I would like to add something. 

In my case, opening a second panel results in unpredictable save locations. As in: Ctrlr is taking the save location of the last opened panel, even when the second panel is closed before saving original panel. Result: Lots of hard work down the drain.

And yes, I noticed one panel getting global variables, if.. else statements and more from another panel. That's probably the same problem you described Goodweather.

Doesn't matter whether it are two tabs in one Ctrlr instance or two separate instances of Ctrlr. Double clicking a panel file on the desktop results in opening Ctrlr with an "already or earlier opened" panel another Ctrlr instance "and" the just double-clicked panel. And we're back at the saving a panel problem.

So, yes, seems like a global Ctrlr problem, unaware of unique ID's or separate instances.

**Reply by dobo365** [reply]
Thx for sharing the concern ;-)
About the Save, you just need to know the steps to avoid frustration and I experienced it as well. Hahaha...
**The trick is to first do a Save As... to a new file name directly followed by a Save**
If you are doing that you are safe.

**Comment by dnaldoog** [comment]
You can fix this by  using the source Luke
<pre>
<code>
x=function(mod,value,source)
   if source == 4 then
          uitls.warnWindow("You made it",""please do something!")
    end
end
</code>
</pre>

The only problem is you will not be able to run this function from another lua function, because the parameter to the sending function will send 5 or 6, so the call  will naturally be blocked!

:)
[Popups bug panel_0_9_added source to block_2021-03-03_21-32.zip](https://github.com/RomanKubiak/ctrlr/files/6075979/Popups.bug.panel_0_9_added.source.to.block_2021-03-03_21-32.zip)

**Reply by dobo365** [reply]
Hmmm... Didn't think of using the source param indeed.
For some methods it will be OK because they are only called by the parent mod but I may have some that I'm calling directly. 
Will investigate. Thx for the tip!
**But the most important: do you have an idea of the reason?**

**Reply by dnaldoog** [reply]
Well for the longest time I was wondering what that parameter, ‘source’ was. I thought it  must be very important for Atom to have included it in any function that reacts to ‘changes in value’ callbacks. You don’t see it in the mouseDown functions. 

Atom must have encountered that same problem.

Right now the only solution I can think of is you might have to have a completely separate message/function that is fired by lua and not the user. 

So in that function you would have to have:

<pre><code>
if source == 5 or source == 6 then
utils .warnWindow(“”,””)
end 
</code></pre>
 
 Doubling up like that could be a drag if the method is very large.

I think I wrote to you a long time ago when we were discussing this issue that my JD990 panel had all of these functions in it that weren’t getting triggered when the panel loaded, I never worked out why they weren’t triggered like this. 

I have tried setting a variable from ‘true’  to ‘false’  on load by using a timer, which is something I think you often do and that sometimes works really well - other times doesn’t work; I still don’t know why. Maybe it has something to do with panels that have a lot of code in them? But if you do a small test panel like the one you did it seems that those functions are exposed / fired off when you load the panel.

**Comment by dnaldoog** [comment]
It seems like you can block functions at boot by adding

<pre><code>
x=function(mod,value,source)
if source ~= 6 then
-- do stuff
end
end -- callback for modulator named myMod1
</pre></code>

If you change another modulator from a second one:

<pre><code>
y=function(mod,value,source)
if source ~= 6 then
x(mod,value,source)
end
end
</pre></code>

then function x() will receive '4' as source. But you could change this to any arbitrary number I suppose.

<pre><code>
y=function(mod,value,source)
if source ~= 6 then
x(nil,value,100)
end
end
</pre></code>

If  the user changes the modulator  - its own callback _Called When the Modulator Changes_  it will pass  '4' to itself. Of course. So instead of source ~=6, you could  do if source==4 if this function is never called except when the user changes it. 

<pre><code>
x=function(mod,value,source)
if source ==4  then
-- do stuff
end
end -- callback for modulator named myMod1
</pre></code>

If you do :

<pre><code>
z=function(mod,value,source)
if source ~= 6 then
panel:getModulatorByName("myMod1"):setValue(value,true)
end
end
</pre></code>

then function x() (attached to myMod1)will receive '5' as source.

**Reply by dobo365** [reply]
Will make different try outs later today or tomorrow but I have good hopes.
I have used the source parameter for other purposes but never thought about using it for this blocking.
And you are right when saying that there must be some "competition" inside Ctrlr when opening panels as indeed I have the issue with some panels and not with others even if they are built with the same philosophy.
If you don't know (or for others), here is the list of the source integer meaning:

From: /Source/Lua/CtrlrLuaModulator.cpp
value(“initialValue”, 0),
value(“changedByHost”, 1),
value(“changedByMidiIn”, 2),
value(“changedByMidiController”, 3),
value(“changedByGUI”, 4),
value(“changedByLua”, 5),
value(“changedByProgram”, 6),
value(“changedByLink”, 7),
value(“changeByUnknown”, 8)

**Reply by Godlike-Productions** [reply]
Here's the original discussion of this at ctrlr forums - I came across it searching for something else.

https://ctrlr.org/forums/topic/build-159/

**Comment by dnaldoog** [comment]
I was looking for that list - thank you. So you could set source at -1 or anything over 9 for custom purposes.
_changedByProgram_ '6' must be at boot time? _changedByMidiIn_ might be interesting '2'.

There is some deeper issue, because I could have sworn a test panel I made showed a _utils.warnWindow()_   using the above logic, but then I couldn't duplicate it.

That JD-990 panel confuses me because I never did any of this and it never runs functions at startup when it should. There must be some timeout or something, so panels with a large number of methods don't do this, but a small panel does. Something like that. Will be interesting to hear about your test results. If this really works well, one could do away with all those bootstrap functions and flag variables set with timers etc.

**Reply by dobo365** [reply]
Yes, if the _source_ parameter is listened to at boot time / panel opening. I'll let yo know the results of my tests.

And, FYI, I have been using the source=2 several times to have the panel showing the right tab based on what the user is doing on the synth. Turning a knob on the synth sends a CC (or NRPN or sysex) to the modulator that is triggering its change method. The first thing in the change method is to switch to the tab where that modulator is lying.

**Comment by dnaldoog** [comment]
Ah so source ==2 is what comes in when the user  changes a knob on the
synth? That’s going to be very useful because then we can change functions
depending on that source parameter different To when the user changes a
lknob on the panel. Excellent.

**Comment by Godlike-Productions** [comment]
Oh, this is such a useful feature that I knew nothing about.

**Reply by dobo365** [reply]
So many things to read, learn and experiment in Ctrlr ;-)
Just need to be willing to spend time on that... Some do, many don't...

**Comment by dobo365** [comment]
OK. I made a few tests and need to do more (esp. with VST instances).
Current status is mixed and can drives you nuts... but OK I learned to be patient and gentle with Ctrlr. Hahahaha ;-)

- the panel I gave you and posted above doesn't give me anymore errors; even with no change. Don't understand this at all...
- I had another less cleaned version that gave the error so I used that one. 
- source=6 is indeed used at panel start and source=4 is used when you press the button in the panel
- as is, I can see that when you open the second panel, each method is called 2 times. I guess once for the first panel and once for the second one
- when adding the source=6 protection, the second panel opens without the popup issue; **Great!!!**
- when clicking on a button in the second panel (Rename for example), the method is called 2 times and the popup appears 2 times, once for the new panel and once for the second one

Further tests to be done:
- different panels but with the same philosophy and method names. I really hope this will work because will make copy/paste of code between panels easier 
- same panel in VST on 2 different tracks

**Reply by dnaldoog** [reply]
Here is a test panel I made that also tests for <code>source==2</code>. You can change the CC number for the uiSlider labelled "incoming", but I found I had to close the panel and reopen it for that to work. 

- When this panel is loaded <code>source = 6</code> is generated. 
- When another modulator modifies another modulator <code>source = 5</code> is generated. 
- When the user changes the value on a modulator <code>source = 4</code> is generated

![testSource](https://user-images.githubusercontent.com/76831683/110118082-266fef80-7dfd-11eb-98da-834fe10b7f42.png)


[source parameter_1_0_0_2021-03-05_21-49.zip](https://github.com/RomanKubiak/ctrlr/files/6090582/source.parameter_1_0_0_2021-03-05_21-49.zip)

**Comment by dobo365** [comment, marked as answer]
I have now done further tests and have an happy face thanks to you, @dnaldoog !
So:
- it is needed to add a test for source==6 and source==1 in the beginning of all methods containing calls to utils.xxx functions
On my panels I have such code:
```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
Load_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action while loading a program or if the panel is in bootstrap or program states
	if bLoadingProgram or not isPanelReady() or source==6 or source==1 then
		return
	end	
...
end
```
- I can now open different panels in parallel
- I can now open the same panel several times either in standalone or in VST
- In VST it is now possible to save DAW presets and replaced them without the multiple popups issue

I'm now adapting all my panels one by one including their manual.
Many thanks again!!!

**Reply by dnaldoog** [reply]
Great to hear Goodweather,

I didn't know about <code>source==1</code> -- will look into that - thanks.
Looks like I might have to go back and change all my panels too!

Here is a suggested function that all this could be wrapped in:

<pre><code>
panelLoadBlock=function(src)
local set = {
  [0] = false, --value(“initialValue”, 0),
  true, --value(“changedByHost”, 1),
  false, --value(“changedByMidiIn”, 2),
  false, --value(“changedByMidiController”, 3),
  false, --value(“changedByGUI”, 4),
  false, --value(“changedByLua”, 5),
  true, --value(“changedByProgram”, 6),
  false,  --value(“changedByLink”, 7),
  false, --value(“changeByUnknown”, 8)
  } -- false allows function to be run 
return set[src]
end -- function
</pre></code>

So, any function that needs to be blocked or muted when the panel loads would have this code at the beginning of the function.

<pre><code>
myFunctionToBlock=function(mod,value,source)
if panelLoadBlock(source) then return end
utils.warnWindow("Hello","World")
end -- my function
</pre></code>

**Reply by dobo365** [reply]
Yes of course with a function is works as well.

I don't know what source == 1 really mean. I thought it was coming when in VST mode (the DAW is the host) but if I'm not wrong I got it while in standalone.

What you must also know is that opening a second panel (different or not from the first one) was launching the methods from the first one. So, this helped me tracked the source values thanks to the console.
I had also to do the opposite to be sure (when opening panel2 after panel1, you need to try panel1 after panel2) to not forget some methods.

I didn't put the source limitation in other methods. Maybe they are also fired but that this is not visible.
Will try that by putting a simple console(tostring(source)) statement in some of them as I did for my main tests.

**Comment by dobo365** [comment]
Some further important info about this topic!

I'm re-using similar _Called when modulator value change_ methods with the same name in all my panels.
I noticed that when you move one modulator in one panel, calling the OnChange method is done with source=4 but then the same method is also launched in the other panel but with source=1
What a mess... I'll report that as a bug.

**Reply by Godlike-Productions** [reply]
How are you using this?  I was going to put a call in isPanelReady to the LoadBlock, but I'm not sure if this is the best way to approach it.  I want to be able to run multiple instances of the same plugin without them interfering with each other.

**Comment by shooking** [comment]
So a neat "old" program I like is snavigator. Been using this since mid 1990s to take code apart.
Let's take an example I had (that you directed me to this thread).
I have a UI slide to set a patch number.
And when this changed, I send the midi from a LUA method.

When I search for "calledByLua" I get a list

![2021-05-18-141115_1920x1080_scrot](https://user-images.githubusercontent.com/22934103/118658572-6d5a6800-b7e4-11eb-99a6-56ffa98f3a6d.png)

So I view some of these instances (you simply click in the grep-editor line of interest and the code snippet is presented)

![2021-05-18-141244_1920x1080_scrot](https://user-images.githubusercontent.com/22934103/118658721-8e22bd80-b7e4-11eb-9313-e3c6c609716f.png)

And finally
![2021-05-18-141359_1920x1080_scrot](https://user-images.githubusercontent.com/22934103/118658792-a1358d80-b7e4-11eb-8406-7975d7797eb5.png)

In the case of a Slider, the CtrlrModulatorValue is being called.
snavigator allows me to find the polymorph for CtrlrSlide
```
void CtrlrSlider::sliderValueChanged (Slider* sliderThatWasMoved)
{
    if (sliderThatWasMoved == &ctrlrSlider)
    {
		if ((bool)owner.getOwnerPanel().getEditor()->getProperty(Ids::uiPanelEditMode) == true)
			return;

		setComponentValue (ctrlrSlider.getValue(), true);
    }
}
```
And so at that level the setCompomentValue is sending true for 2nd param.
What means true?
```
void CtrlrSlider::setComponentValue (const double newValue, const bool sendChangeMessage)
{
	ctrlrSlider.setValue (newValue, dontSendNotification);
	if (sendChangeMessage)
	{
		owner.getProcessor().setValueGeneric (CtrlrModulatorValue(newValue,CtrlrModulatorValue::changedByGUI));
	}
}
```
Seems to be using the trick you explained - changedByGUI == 4

So it looks like IF some of these additional parameters are not exposed (ie if I cannot sendChangeMessage = false) then I need to look at breaking the chain another way.

I think what is happening:
move UI slide -> set value in UI, then hardcoded to sendChangeMessage ie modify the mod value as changedByGUI.
Likely this triggers my callback ... and that sends MIDI to the pedal, but is also intercepted in the handleMidiMessage.
Now this seems to be where the source, as you highlighted, is coming into the picture.
Because I see this (I just sent it!) and then try to update the modulator - maybe changedByLua?
And this sets the value ... which seems to think it is changedByGUI ... this we now do!!

I feel a debugging session coming on.
